consensus/core/src/synchronizer.rs

```diff
@@ -649,6 +692,111 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
+
+                for (authority_index, _authority) in context.committee.authorities() {
+                    if authority_index != context.own_index {
+                        results.push(fetch_own_block(authority_index, Duration::from_millis(0)));
+                    }
+                }
+
+                // Gather the results but wait to timeout as well
+                let timer = sleep_until(Instant::now() + context.parameters.sync_last_proposed_block_timeout);
+
+                tokio::pin!(timer);
+
+                // Get the highest of all the results
+                let mut total_stake = 0;
+                let mut highest_round = 0;
+                'main_loop: loop {
+                    tokio::select! {
+                        Some((result, authority_index)) = results.next() => {
+                            match result {
+                                Ok(result) => {
+                                    let mut blocks = Vec::new();
+                                    for serialized_block in result {
+                                        let Ok(signed_block) = bcs::from_bytes(&serialized_block) else {
+                                            warn!("Malformed block received from peer {authority_index} while fetching own last block, skipping.");
+                                            continue 'main_loop;
+                                        };
+
+                                        if let Err(e) = block_verifier.verify(&signed_block) {
+                                            context
+                                                .metrics
+                                                .node_metrics
+                                                .invalid_blocks
+                                                .with_label_values(&[&signed_block.author().to_string(), "synchronizer_own_block"])
+                                                .inc();
+                                            warn!("Invalid block received from {}: {}", authority_index, e);
+                                            continue 'main_loop;
+                                        }
+
+                                        let verified_block = VerifiedBlock::new_verified(signed_block, serialized_block);
+                                        if verified_block.author() != context.own_index {
+                                            continue 'main_loop;
+                                        }
+                                        blocks.push(verified_block);
+                                    }
+
+                                    let max_round = blocks.into_iter().map(|b|b.round()).max().unwrap_or(0);
+                                    highest_round = highest_round.max(max_round);
+
+                                    total_stake += context.committee.stake(authority_index);
+
+                                    if results.is_empty() {
+                                        break;
```

Comment:
Is this `main_loop as well, or should we remove the loop label?

Reply:
This is main loop yes, but why removing the label? We use the label in order to continue to next peer when we get invalid results from the current we process.